File: problem/0007.py
class Solution:
    def reverse(self, x: int) -> int:
        # # python3 number is 64-bits
        # # method 1
        # isPostive = x >= 0
        # x = abs(x)
        # result = 0
        # while x != 0:
        #     pop = x % 10
        #     result = result * 10 + pop
        #     # check range
        #     if (isPostive and result > 2 ** 31 - 1) or (not isPostive and result > 2 ** 31):
        #         return 0
        #     x = x // 10
        # return result if isPostive else -result
        
        # method 2, reverse as string
        reverse = int(str(x * -1)[::-1]) * -1 if x < 0 else int(str(x)[::-1])
        return 0 if (reverse > 2 ** 31 - 1 or reverse < -2 ** 31) else reverse

File: problem/0007.java
class Solution {
    public int reverse(int x) {
//         // method 1, use array to store and process
//         int DIGITS_SIZE = 10;
//         boolean isPostive = x >= 0;
//         int[] nums = new int[DIGITS_SIZE];
//         int count = 0;
//         while (x != 0 && count < DIGITS_SIZE) {
//             nums[count++] = x % 10;
//             x /= 10;
//         }
//         int result = 0;
//         for (int i = count - 1; i >= 0; i--) {
//             // check int overflow
//             if ((isPostive && (nums[i] > (Integer.MAX_VALUE - result) / Math.pow(10, count - i - 1))) ||
//                (!isPostive && (nums[i] < (Integer.MIN_VALUE - result) / Math.pow(10, count - i - 1)))) return 0;
//             result += nums[i] * Math.pow(10, count - i - 1);
//         }
//         return result;

//         // method 2, offical soultion, process in every step
//         int rev = 0;
//         while (x != 0) {
//             int pop = x % 10;
//             x /= 10;
//             if (rev > Integer.MAX_VALUE/10 || (rev == Integer.MAX_VALUE / 10 && pop > 7)) return 0;
//             if (rev < Integer.MIN_VALUE/10 || (rev == Integer.MIN_VALUE / 10 && pop < -8)) return 0;
//             rev = rev * 10 + pop;
//         }
//         return rev;

        // method 3, 
        // https://leetcode.com/problems/reverse-integer/discuss/4060/My-accepted-15-lines-of-code-for-Java
        int result = 0;
        while (x != 0)
        {
            int tail = x % 10;
            int newResult = result * 10 + tail;
            if ((newResult - tail) / 10 != result) return 0;
            result = newResult;
            x = x / 10;
        }
        return result;
    }
}